src/js/common/parser.js

```diff
@@ -0,0 +1,188 @@
+/*
+
+    A parser that resolves `$ref` references in JSON Schema definitions and provides the ability to look up the
+    dereferenced definition content based on a "dot notation" path within a deep object.  See the documentation for
+    details:
+
+    https://github.com/the-t-in-rtf/gpii-json-schema/blob/GPII-1336/docs/parser.md
+
+*/
+"use strict";
+var fluid = fluid || require("infusion");
+var gpii  = fluid.registerNamespace("gpii");
+
+var $RefParser = $RefParser ? $RefParser : require("json-schema-ref-parser");
+var jsonpointer = jsonpointer ? jsonpointer : require("jsonpointer.js");
+
+fluid.registerNamespace("gpii.schema.parser");
+
+// Dereference all `$ref` links for a single schema.  See the documentation for details:
+//
+// https://github.com/the-t-in-rtf/gpii-json-schema/blob/GPII-1336/docs/parser.md#gpiischemaparserdereferenceschemathat-schemapath-schemakey
+//
+gpii.schema.parser.dereferenceSchema = function (that, schemaPath, schemaKey) {
+    var parser = new $RefParser(); // jshint ignore:line
+    var pathOrUri = fluid.stringTemplate(that.options.uriTemplate, { schemaPath: schemaPath, schemaKey: schemaKey});
+    var promise = fluid.promise();
+    parser.dereference(pathOrUri, that.options.parserOptions, gpii.schema.parser.getParserCallback(that, schemaKey, promise));
+    return promise;
+};
+
+// Wrap the normal callback used by the parser in a `fluid.promise`.
+gpii.schema.parser.getParserCallback = function (that, schemaKey, promise) {
+    return function (error, schema) {
+        if (error) {
+            promise.reject(error);
+        }
+        else {
+            that.dereferencedSchemas[schemaKey] = schema;
+            promise.resolve(schema);
+        }
+    };
+};
+
+// Look up the full dereferenced definition for a single field.  See the documentation for details:
+//
+// https://github.com/the-t-in-rtf/gpii-json-schema/blob/GPII-1336/docs/parser.md#gpiischemaparserresolvejsonpointerthat-schemakey-schemafieldpath
+//
+gpii.schema.parser.resolveJsonPointer = function (that, schemaKey, jsonPointer) {
+    var dereferencedSchema = that.dereferencedSchemas[schemaKey];
+    return jsonpointer.get(dereferencedSchema, jsonPointer);
+};
+
+
+// Look up the error metadata for a single field. See the documentation for details:
```

Comment:
Sadly I fear that until we have a docs site set up, the docs comments here will need to duplicate the material which is in the .md file. This is what we do in core Infusion as well - readers of the code need to be able to read it in a self-contained way. Also, you already have a cut and paste burden since all of these links will change once the code is committed.